Show a link to the commit in the result details

MaisiKoleni opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, editor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Client: I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-testing/).
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all newly inserted strings into English and German.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).

### Motivation and Context
Currently, students cannot see what result belongs to which commit, especially in the code editor or exercise details. Usually, this is just the latest commit in the project, but it can be the case that either a corresponding result is not visible for them or has not been generated, see #3976.

### Description
Add a link to the top of the result details modal in case of programming exercises. We reuse the logic from the participation submission component here, and moved the commit link creation logic into the programming exercise utilities to avoid duplication.

In case the exercise is not available, we just show the commit in plain text. If the submission is not available, we do not display the note about the linked commit at all, as it is not possible then.

### Steps for Testing
1. Work with programming exercises: commit and wait for results. Some suggestions: try template, solution and test repository commits as well as regular student working in the code editor. It is also OK to use old results here.
2. Open the result details (click the result string)
3. You should see a message at the top of the feedback modal that links to a commit as shown in the screenshot.
4. Check that the link works.

*It might also be useful to test different locations where results can appear. It does currently not work for the assessment code editor.*

You can also do a quick check that it does not appear for other exercise results.

### Review Progress
<!-- Each Pull Request should be reviewed by at least two other developers. The code as well as the functionality (= manual test) needs to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->

- Code Review
  - [x] Review 1
  - [x] Review 2
- Manual Tests
  - [x] Test 1
  - [x] Test 2

### Test Coverage
* result-detail.component.ts 95.36%
  ![image](https://user-images.githubusercontent.com/11130248/132477583-7d708e31-2823-4e70-bbb6-7739dab5e41b.png)
* programming-exercise.utils.ts → full branch coverage of `createCommitUrl`
  ![image](https://user-images.githubusercontent.com/11130248/133668475-865e289e-f5b5-40ea-baa3-b0d64279cfe3.png)

### Screenshots
![image](https://user-images.githubusercontent.com/11130248/132369406-ac13874f-7739-45a0-9077-4faa166a8821.png)

julian-christl commented on src/main/webapp/app/entities/participation/participation.model.ts:
```diff
@@ -49,7 +49,12 @@ export abstract class Participation implements BaseEntity {
+ *
+ * This is needed as different participation types may use different exercise attributes (e.g. not 'exercise' but 'programmingExercise').
+ */
+export const getExercise = (participation: Participation | undefined): Exercise | undefined => {
```
I don't understand why this should be undefined-able. If I ask for the exercise of undefined it should be trivial that this is not possible to retrieve -> undefined should be captured before accessing this function.

MaisiKoleni replied:
I thought it might simplify handling possibly undefined values, but you are right, I can revert this. I would also remove the `if` then.

julian-christl commented on src/main/webapp/app/exercises/shared/result/result-detail.component.html:
```diff
@@ -75,6 +75,14 @@ <h4 jhiTranslate="artemisApp.result.buildError">Build Error:</h4>
+ -->
+    <p *ngIf="exerciseType === ExerciseType.PROGRAMMING && result.submission">
+        <span>{{ 'artemisApp.result.linkedCommit' | artemisTranslate }}</span>
+        <a *ngIf="getCommitUrl(); else commitWithoutLink" href="{{ getCommitUrl() }}" target="_blank" rel="noopener noreferrer">{{ getCommitHash().substr(0, 11) }}</a>
```
I think these function calls could be pre-calculated with an attribute (Client guidelines Section 0)

MaisiKoleni replied:
Yes, we can do that, it is quite unlikely they will change.

julian-christl commented on src/main/webapp/app/exercises/shared/participation-submission/participation-submission.component.ts:
```diff
@@ -174,27 +179,7 @@ export class ParticipationSubmissionComponent implements OnInit {
     }
 
     getCommitUrl(submission: ProgrammingSubmission): string | undefined {
```
Return value is `string`.
However, it might make sense to consider returning undefined instead of an empty string in the `createCommitUrl()` and handling the undefined event.

MaisiKoleni replied:
I just extracted the function from this component, but we can change that.

julian-christl commented on src/main/webapp/app/exercises/programming/shared/utils/programming-exercise.utils.ts:
```diff
@@ -25,6 +28,35 @@ export const createBuildPlanUrl = (template: string, projectKey: string, buildPl
+    participation: Participation | undefined,
+    submission: ProgrammingSubmission,
+): string => {
+    const projectKeyLowerCase = projectKey?.toLowerCase();
+    let repoSlug: string | undefined = undefined;
+    if (participation?.type === ParticipationType.PROGRAMMING) {
+        const studentParticipation = participation as ProgrammingExerciseStudentParticipation;
+        if (studentParticipation.repositoryUrl) {
+            repoSlug = projectKeyLowerCase + '-' + studentParticipation.participantIdentifier;
+        }
+    } else if (participation?.type === ParticipationType.TEMPLATE) {
+        // In case of a test submisson, we need to use the test repository
+        repoSlug = projectKeyLowerCase + (submission?.type === SubmissionType.TEST ? '-tests' : '-exercise');
+    } else if (participation?.type === ParticipationType.SOLUTION) {
+        // In case of a test submisson, we need to use the test repository
+        repoSlug = projectKeyLowerCase + (submission?.type === SubmissionType.TEST ? '-tests' : '-solution');
+    }
+    if (repoSlug && template && projectKeyLowerCase) {
+        return template
+            .replace('{projectKey}', projectKeyLowerCase)
+            .replace('{repoSlug}', repoSlug)
+            .replace('{commitHash}', submission.commitHash ?? '');
+    }
+    return '';
```
```suggestion
    const projectKeyLowerCase = projectKey?.toLowerCase();
    let repoSlug: string | undefined = undefined;
    if (participation?.type === ParticipationType.PROGRAMMING) {
        const studentParticipation = participation as ProgrammingExerciseStudentParticipation;
        if (studentParticipation.repositoryUrl) {
            repoSlug = studentParticipation.participantIdentifier;
        }
    } else if (participation?.type === ParticipationType.TEMPLATE) {
        // In case of a test submisson, we need to use the test repository
        repoSlug = (submission?.type === SubmissionType.TEST ? 'tests' : 'exercise');
    } else if (participation?.type === ParticipationType.SOLUTION) {
        // In case of a test submisson, we need to use the test repository
        repoSlug = (submission?.type === SubmissionType.TEST ? 'tests' : 'solution');
    }
    if (repoSlug && template && projectKeyLowerCase) {
        return projectKeyLowerCase + '-' + template
            .replace('{projectKey}', projectKeyLowerCase)
            .replace('{repoSlug}', repoSlug)
            .replace('{commitHash}', submission.commitHash ?? '');
    }
    return '';
```
I think this is equivalent. However, this seems a bit odd since you add the projectKey before the template and replace it in the template. Could the template here be changed to fit it better?

MaisiKoleni replied:
I think I get what you mean, but this is not equivalent. You probably want `projectKeyLowerCase + '-' + repoSlug` in the replacement. I would then rename the variable, as it is no longer the repo slug but the postfix.

julian-christl replied:
Ah, yes. You are correct. My bad. But good change!

julian-christl commented on src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.ts:
```diff
@@ -88,12 +88,16 @@ export class ProgrammingExerciseDetailComponent implements OnInit, OnDestroy {
                         this.programmingExercise.templateParticipation.results = [latestTemplateResult];
                     }
+                    // This is needed to access the exercise in the result details
+                    this.programmingExercise.templateParticipation.programmingExercise = this.programmingExercise;
```
I suggest using Angular's `@Input` instead. Manipulating the model is quite error-prone. => Then also revert the other file in that regard

MaisiKoleni replied:
I don't see that adding another input is less error-prone than fixing the association here. Note that this is for some reason only a problem for template and solution, and I assume the added lines are missing. We manipulate the model in many places (and need to), including the lines above and below these two. With the input, we would effectively have two exercises in the context.

julian-christl replied:
So the reason I prefer the Input here is that we don't have a dependency chain that could have been manipulated at another place and therefore break anything because we directly supply the exercise.

Might this also be something worth discussing? 👀 😅

MaisiKoleni replied:
Well, at least the components changed here are read-only, they don't do updates on the server. The issue would be the same for the submission of the result and the participation, both of which are also indirectly referenced in the result component. And we have many, many more in the client. We can discuss this further in the team, but I see more problems arising from that than it solves. At least for now, I couldn't name any actual problem we would solve. But it is quite common in Artemis to have attributes available sometimes, and sometimes not. Refactoring that is a lot of work, I am not sure how beneficial that would turn out eventually / how realistic that is.

I would regard it at least out of scope for this PR.

Some additional info on what I've done:
- I did a bit of web search, but could not find that any approach was considered dangerous / preferred and why, but my searches may have missed something.
- I tried checking the only two possible "update" actions, which are deleting a result or submission.
  Both currently cause different bugs in Artemis for template and solution. Result deletion simply is not executed and the modal "hags", submission deletion causes a class cast exception on the server. Looks like no-one every noticed that they were broken.

julian-christl replied:
I see. Thanks for taking a look into it.
I expected the Input to have a proper check but that is not the case as we can see from the issue with the if clause.
So yeah, probably currently not worth digging into too much.

Could you maybe create issues for the two bugs you've found so they are documented properly?

MaisiKoleni replied:
Done, see #4008 and #4009.

julian-christl commented on src/test/javascript/spec/component/shared/result-detail.component.spec.ts:
```diff
@@ -163,24 +177,49 @@ describe('ResultDetailComponent', () => {
             });
     });
 
+    it('should generate commit link for programming exercise result with submission, participation and exercise', () => {
```
Could you also add tests for the edge cases (n/a or empty string/undefined)?

MaisiKoleni replied:
I can look it up, but the commit link generating function is already tested somewhere else (where it originally came from).

julian-christl replied:
But didn't you move the test as well?

MaisiKoleni replied:
No, the test is entirely new, to check this component specifically.

julian-christl replied:
Took another closer look, now I get it, sorry.

If you move the other one, I think it makes sense to add those two small tests though. Then this function is properly tested.

julian-christl reviewed (changes requested):
Nice addition, got some questions.

julian-christl commented on src/main/webapp/app/entities/participation/participation.model.ts:
```diff
@@ -54,7 +54,7 @@ export abstract class Participation implements BaseEntity {
  */
-export const getExercise = (participation: Participation | undefined): Exercise | undefined => {
+export const getExercise = (participation: Participation): Exercise | undefined => {
     if (participation) {
```
You forgot to remove the if clause here 😅

MaisiKoleni replied:
Sadly, this is needed, as the client tests otherwise fail. There seems to be some code relying on the existence of this `if`.

julian-christl replied:
What client tests fail? Because how can participation be here not truthy? I'm confused there :D

MaisiKoleni replied:
Also, see aa8379e1124c922509bca12aa44d0bdcc67d1d0f and https://github.com/ls1intum/Artemis/runs/3599962633#step:5:1488. While this is one error, it is the only one the tests found, and now I am not sure how much I can trust the type checking / our code base here.

julian-christl replied:
I know what the issue is: In `result.component.ts` `getExercise()` gets called if the participation was not set.
participation is a non optional Participation and therefore the typecheck resolves. But in the HTML apparently no participation gets set => getExercise gets called with undefined.

So you could add a undefined check in the `evaluateTemplateStatus()`. Then removing the if clause should work.

MaisiKoleni replied:
Looking through the code, there are more occurrences that would need to be checked and more usages of such "unnecessary" if's. In short, I don't know where we need to add if's to remove this single one. I would suggest letting it be for now because it works, and I don't know whether my refactored version will do so.

julian-christl replied:
ok

julian-christl reviewed (changes requested):
Just one small issue

julian-christl commented on src/test/javascript/spec/component/participation-submission/participation-submission.component.spec.ts:
```diff
@@ -175,7 +175,7 @@ describe('ParticipationSubmissionComponent', () => {
         comp.participation = participation;
         comp.exercise = exercise;
-        expect(comp.getCommitUrl(submission)).to.be.empty;
+        expect(comp.getCommitUrl(submission)).to.be.undefined;
```
I think it makes more sense to move this test and test the createCommitUrl() instead of the getter since the getter ist atm just a wrapper for the createCommitUrl()

FlorianGlombik reviewed (changes requested):
The commit hash should always be a clickable link, which is not the case when being opened from the problem statement and not the code editor.
![image](https://user-images.githubusercontent.com/63976129/133406541-b01c8b95-e47c-4387-a44c-d4545a50d487.png)

This Screencast visualizes the description:
![CommitHashOnlyClickableFromCodeEditor](https://user-images.githubusercontent.com/63976129/133406921-1e2ab5f2-da7f-4b84-9341-dc159e2151e6.gif)

_(Tested it on ts1 with an existing exercise but new participation.)_

MaisiKoleni commented:
> The commit hash should always be a clickable link, which is not the case when being opened from the problem statement and not the code editor.
> *<< IMAGE >>*
> 
> This Screencast visualizes the description:
> *<< SCREEN RECORDING >>*
> 
> _(Tested it on ts1 with an existing exercise but new participation.)_

Should be fixed with adc1605 now.

MaisiKoleni commented:
@julian-christl I now moved one test and added many more tests for that specific function. I left one in each component to have the function there covered as well and check that it works.

julian-christl reviewed (dismissed):
Looks good now, thanks for the changes and the effort with the tests!

MerlinMehmed reviewed (dismissed):
Tested in a testing session and works as expected.

I think that as a follow up the date and time of the commit can be added.

rs-maker reviewed (approved):
worked fine in testing session on ts1.

Malyuk-A reviewed (approved):
Reviewed during testing session. Code looks good.

MaisiKoleni commented:
Should be ready to merge, both result PRs also work well in combination:
![image](https://user-images.githubusercontent.com/11130248/133824709-c7118988-b1e0-46bc-b3dc-6c1f07d5bf4c.png)

julian-christl reviewed (approved):
Reapprove code.

rs-maker reviewed (approved):
re approve.

Malyuk-A reviewed (approved):
re-approve

MerlinMehmed reviewed (approved):
It still works. Reapprove!
